src/com/sun/jna/win32/W32StringUtil.java

```diff
@@ -0,0 +1,390 @@
+import com.sun.jna.Pointer;
+
+/**
+ * Native string methods for Windows.
+ *
+ * @author Torbj&ouml;rn Svensson, azoff[at]svenskalinuxforeningen.se
+ */
+public abstract class W32StringUtil {
+    /**
+     * Returns true for unicode API.
+     *
+     * @return True for unicode API
+     */
+    public static boolean isAPITypeWide() {
+        return W32APITypeMapper.DEFAULT == W32APITypeMapper.UNICODE;
+    }
+
+    /**
+     * Return the width of TCHAR in C
+     *
+     * @return The width of TCHAR in C
+     */
+    public static int getCharWidth() {
+        if (isAPITypeWide()) {
+            return Native.WCHAR_SIZE;
+        }
+        return 1;
+    }
+
+    /**
+     * Convert buffer to string
+     *
+     * @param pointer The buffer
+     * @return The string
+     * @see #toString(Pointer)
+     * @see #toString(Pointer, int)
+     * @see #toString(Pointer, int, int)
+     */
+    public static String toString(Pointer pointer) {
+        return toString(pointer, 0);
+    }
+
+    /**
+     * Convert buffer to string
+     *
+     * @param pointer The buffer
+     * @param offsetInCharacters The offset in characters
+     * @return The string
+     * @see #toString(Pointer)
+     * @see #toString(Pointer, int, int)
+     */
+    public static String toString(Pointer pointer, int offsetInCharacters) {
+        if (isAPITypeWide()) {
+            return pointer.getWideString(offsetInCharacters * getCharWidth());
+        } else {
+            return pointer.getString(offsetInCharacters);
+        }
+    }
+
+    /**
+     * Convert buffer to string
+     *
+     * @param pointer The buffer
+     * @param offsetInCharacters The offset in characters
+     * @param lengthInCharacters The length of the string in characters
+     * @return The string
+     * @see #toString(Pointer)
+     * @see #toString(Pointer, int)
+     */
+    public static String toString(Pointer pointer, int offsetInCharacters, int lengthInCharacters) {
+        byte[] bytes = pointer.getByteArray(offsetInCharacters * getCharWidth(), lengthInCharacters * getCharWidth());
+        return toString(bytes, 0);
+    }
+
+    /**
+     * Convert buffer to string
+     *
+     * @param buffer The buffer
+     * @return The string
+     * @see #toString(byte[], long)
+     */
+    public static String toString(byte[] buffer) {
+        return toString(buffer, 0);
+    }
+
+    /**
+     * Convert buffer to string
+     *
+     * @param buffer The buffer
+     * @param offsetInCharacters The offset in characters
+     * @return The string
+     * @see #toString(byte[])
+     */
+    public static String toString(byte[] buffer, long offsetInCharacters) {
+        Memory pointer = new Memory(buffer.length + 1 * getCharWidth());
+        pointer.write(0, buffer, 0, buffer.length);
+
+        // Ensure null-termination
+        byte[] zeros = new byte[getCharWidth()];
+        Arrays.fill(zeros, (byte)0);
+        pointer.write(buffer.length, zeros, 0, zeros.length);
```

Comment:
`Memory#clear`

Reply:
While that would do the trick, it would mean writing a hole bunch of nulls. I can switch to clear if you think that's better than writing one or two bytes.

Reply:
It is a native call backed by `memset` and already used in `com.sun.jna.Structure.AutoAllocated` (the memory backing structures). So it is most probably already in a hotpath.

Reply:
Still, a call to memset is more expensive than writing 2 bytes, right?

Reply:
I have doubts that there is a huge difference, but as written, clearing memory is already done in the Structure call path and the most probably the domineering factor is most probably the transition to native which is needed in both cases.

Reply:
Ok. I guess clear() is cleaner, so sure! I avoided it as I though it was more expensive than writing one or two bytes...